packages/teams-js/src/private/externalAppCommands.ts

```diff
@@ -0,0 +1,208 @@
+import { errorNotSupportedOnPlatform } from '../public/constants';
+import { runtime } from '../public/runtime';
+import { ExternalAppErrorCode } from './constants';
+
+/**
+ * v2 APIs telemetry file: All of APIs in this capability file should send out API version v2 ONLY
+ */
+const externalAppCommandsTelemetryVersionNumber: ApiVersionNumber = ApiVersionNumber.V_2;
+
+/**
+ * @hidden
+ * Namespace to delegate the ActionCommand to the host
+ * @internal
+ * Limited to Microsoft-internal use
+ *
+ * @beta
+ */
+export namespace externalAppCommands {
+  /**
+   * @hidden
+   * The payload of IActionCommandResponse
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export type IActionCommandResponse = ITextActionCommandResponse | ICardActionCommandResponse;
+
+  /**
+   * @hidden
+   * The payload of IBaseActionCommandResponse
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export interface IBaseActionCommandResponse {
+    taskModuleClosedReason: TaskModuleClosedReason;
+  }
+
+  /**
+   * @hidden
+   * The text result type
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export interface ITextActionCommandResponse extends IBaseActionCommandResponse {
+    resultType: 'text';
+    text: string | undefined;
+  }
+
+  /**
+   * @hidden
+   * The card result type
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export interface ICardActionCommandResponse extends IBaseActionCommandResponse {
+    resultType: 'card';
+    attachmentLayout: AttachmentLayout;
+    attachments: QueryMessageExtensionAttachment[];
+  }
+
+  /**
+   * @hidden
+   * The result type for the ActionCommandResultType
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export type ActionCommandResultType = 'card' | 'text';
+
+  /**
+   * @hidden
+   * The attachment layout type for the AttachmentLayout
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export type AttachmentLayout = 'grid' | 'list';
+
+  /**
+   * @hidden
+   * The reason for the TaskModuleClosedReason
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export type TaskModuleClosedReason = 'Done' | 'CancelledByUser';
+
+  /**
+   * @hidden
+   * The data model of QueryMessageExtensionCard
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export type QueryMessageExtensionCard = {
+    contentType: string;
+
+    content: Record<string, unknown>;
+
+    fallbackHtml?: string;
+
+    signature?: string;
+  };
+
+  /**
+   * @hidden
+   * The data model of QueryMessageExtensionAttachment
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export type QueryMessageExtensionAttachment = QueryMessageExtensionCard & {
+    preview?: QueryMessageExtensionCard;
+  };
```

Comment:
This definition of `QueryMessageExtensionAttachment` allows for `preview` to be `undefined`. Isn't that just a `QueryMessageExtensionCard` then? Would the only value of having a new `QueryMessageExtensionAttachment` type be if the `preview` property was actually present (i.e., not `undefined`)?

Reply:
`This definition of QueryMessageExtensionAttachment allows for preview to be undefined. Isn't that just a QueryMessageExtensionCard then?` Yes, it is. When the preview is defined, the type is QueryMessageExtensionAttachment; otherwise the type is QueryMessageExtensionCard. Thanks @lakhveerk for confirming the logic!

Reply:
I updated the comment to make sure well-explained the logic to app developers.